File: Medium/Minimum Spanning Tree/minimum-spanning-tree.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution
{
	public:
	//Function to find sum of weights of edges of the Minimum Spanning Tree.
    int spanningTree(int V, vector<vector<int>> adj[])
    {
        // code here
        vector<int>visited(V,0);
        
        priority_queue<pair<int,int>, vector<pair<int,int>> , greater<pair<int,int>> > pq;
        
        pq.push({0,0});
        int sum=0;
        
        while(!pq.empty())
        {
            pair<int,int> front= pq.top();
            
            int wt = front.first;
            int e = front.second;
            
            pq.pop();
            
            if(visited[e]==1)
            continue;
            sum+=wt;
            visited[e]=true;
            
            
            for(auto x : adj[e])
            {
                int edge = x[0];
                int w = x[1];
                if(visited[edge]==0)
                {
                    pq.push({w,edge});
                }
            }
        }
        return sum;
    }
};

//{ Driver Code Starts.


int main()
{
    int t;
    cin >> t;
    while (t--) {
        int V, E;
        cin >> V >> E;
        vector<vector<int>> adj[V];
        int i=0;
        while (i++<E) {
            int u, v, w;
            cin >> u >> v >> w;
            vector<int> t1,t2;
            t1.push_back(v);
            t1.push_back(w);
            adj[u].push_back(t1);
            t2.push_back(u);
            t2.push_back(w);
            adj[v].push_back(t2);
        }
        
        Solution obj;
    	cout << obj.spanningTree(V, adj) << "\n";
    }

    return 0;
}


// } Driver Code Ends

File: Difficulty: Medium/Minimum Spanning Tree/minimum-spanning-tree.py
#User function Template for python3
from typing import List
import heapq
class Solution:
    
    #Function to find sum of weights of edges of the Minimum Spanning Tree.
    def spanningTree(self, V: int, adj: List[List[int]]) -> int:
        #code here
        visited = [0]*V
        q = []
        heapq.heappush(q,(0,0))
        mst=0
        
        while q:
            wt,node = q[0]
            heapq.heappop(q)
            
            if visited[node]==1:
                continue
            
            visited[node]=1
            mst+=wt
            
            for v,w in adj[node]:
                
                if visited[v]==0:
                    heapq.heappush(q,(w,v))
        
        return mst


#{ 
 # Driver Code Starts
#Initial Template for Python 3
import atexit
import io
import sys
from typing import List

# Contributed by : Nagendra Jha

if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        V, E = map(int, input().strip().split())
        adj = [[] for i in range(V)]
        for i in range(E):
            u, v, w = map(int, input().strip().split())
            adj[u].append([v, w])
            adj[v].append([u, w])
        ob = Solution()

        print(ob.spanningTree(V, adj))
        print("~")

# } Driver Code Ends